### PR TITLE
Add SSHManager support for invoking Windows workers via cmd.exe

### PR DESCRIPTION
**Pull request by mgkuhn**
Distributed.addprocs() now supports four new keyword arguments `shell`,
`ssh`, `env` and `cmdline_cookie`.

Specifying `shell=:wincmd` now makes it possible to start workers on a
Windows machine with an sshd server that invokes `cmd.exe` as the shell
(e.g. Microsoft's OpenSSH port does that by default). Previously
SSHManager only supported ssh connections to a POSIX shell.

Specifying `ssh="/usr/bin/ssh"` makes it possible to specify the ssh
client that SSHManager will use (useful for debugging and where a
custom-version of ssh is required).

The new `env` parameter now allows to pass arbitrary environment
variables to workers.

Specifying `cmdline_cookie=true` is a workaround for an ssh problem
with Windows workers that run older (pre-ConPTY) version of Windows,
Julia, or OpenSSH.

(This is a revised version of PR #30614)

### CODE REVIEW

**Comment by musm** [comment]
LGTM pending https://github.com/JuliaLang/julia/pull/38352#pullrequestreview-534057968

**Review by musm — APPROVED** [review]

**Comment by jebej** [comment]
Is it possible to make this work for when ssh conects to powershell instead of cmd? Maybe by first starting cmd.exe?

**Comment by mgkuhn** [comment]
Adding PowerShell command-invocation support is on my todo list, but will be a future, separate PR. (Proper meta-character escaping in PowerShell sadly can be even more complicated than in `cmd.exe`. Neither has an escaping mechanism as simple and clean as we are used to from the POSIX shell.)

**Comment by jebej** [comment]
So it wouldn't be possible to just start cmd.exe from powershell and then use the work from this PR?

**Comment by musm** [comment]
Looks like this needs a rebase  to remove the cmd.exe escaping functions since we've merged  https://github.com/JuliaLang/julia/pull/38352

**Comment by mgkuhn** [comment]
Rebased after merge of #38352

**Comment by musm** [comment]
Failures look unrelated?

**Pull request merged**
